infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/utils/LogReplicationConfigManager.java

```diff
@@ -142,9 +326,149 @@ private boolean registryTableHasNewEntries() {
+                    throw new RetryNeededException();
+                }
+            }).run();
+        } catch (InterruptedException e) {
+            log.error("Unable to preprocess client configuration tables", e);
+            throw new UnrecoverableCorfuInterruptedError(e);
+        }
+    }
+
+    /**
+     * Callback method which is invoked when the client config listener receives update for new client registration.
+     */
+    public void onNewClientRegister(ReplicationSubscriber subscriber) {
+        if (registeredSubscribers.contains(subscriber)) {
+            log.warn("Client {} with model {} already registered!", subscriber.getClientName(), subscriber.getModel());
+            return;
+        }
+        // TODO (V2): Currently we add a default subscriber for logical group use case instead of listening on client
+        //  registration. Subscriber should be added upon registration after grpc stream for session creation is added.
+        registeredSubscribers.add(subscriber);
+    }
+
+
+    /**
+     * Callback method which is invoked when the client config listener receives update for adding/removing
+     * Sinks from a logical group.
+     *
+     * @param subscriber   Registered replication subscriber that changes the group destination mapping.
+     * @param logicalGroup Logical group that is being modified.
+     * @param destinations The updated list of destinations that this logicalGroup is targeting to.
+     * @return A set of LogReplicationSession that needs a forced snapshot sync
+     */
+    public Set<LogReplicationSession> onGroupDestinationsChange(ReplicationSubscriber subscriber,
+                                                                String logicalGroup,
+                                                                List<String> destinations) {
+        if (!registeredSubscribers.contains(subscriber)) {
+            log.error("Client {} has not registered with replication model {}. Please register first.",
+                    subscriber.getClientName(), subscriber.getModel());
+            return null;
+        }
+
+        Set<String> recordedSinks = groupSinksMap.getOrDefault(logicalGroup, new HashSet<>());
+        Set<String> updatedSinks = new HashSet<>(destinations);
+
+        // The impacted sinks are those who are newly added to this logicalGroup or removed from it. Their config
+        // needs to be generated again to reflect the group destination change.
+        Set<String> impactedSinks = Stream.concat(recordedSinks.stream(), updatedSinks.stream())
+                .filter(sink -> !recordedSinks.contains(sink) || !updatedSinks.contains(sink))
+                .collect(Collectors.toSet());
+        // Update groupSinksMap first and then generate config, as config generation relies on groupSinksMap as well.
+        groupSinksMap.put(logicalGroup, updatedSinks);
+        generateConfig(sessionToConfigMap.keySet().stream()
```

Comment:
we should not generate the config for sinks which are getting removed.

Reply:
Actually it's needed, we need to update the config at Source side for all impacted sessions. It was the forced snapshot sync only triggered for sessions with added groups in this patch.